presto-thrift/src/main/java/com/facebook/presto/thrift/clientproviders/RetryingPrestoClientProvider.java

```diff
@@ -0,0 +1,207 @@
+import com.facebook.presto.thrift.interfaces.client.ThriftTableLayout;
+import com.facebook.presto.thrift.interfaces.client.ThriftTableLayoutResult;
+import com.facebook.presto.thrift.interfaces.client.ThriftTupleDomain;
+import com.facebook.presto.thrift.util.RetryDriver;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.inject.BindingAnnotation;
+import io.airlift.log.Logger;
+import io.airlift.units.Duration;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+import javax.inject.Inject;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+import static java.lang.annotation.ElementType.PARAMETER;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+import static java.util.Objects.requireNonNull;
+
+public class RetryingPrestoClientProvider
+        implements PrestoClientProvider
+{
+    private static final Logger log = Logger.get(RetryingPrestoClientProvider.class);
+    private final PrestoClientProvider original;
+    private final RetryDriver retry;
+
+    @Inject
+    public RetryingPrestoClientProvider(@NonRetrying PrestoClientProvider original)
+    {
+        this.original = requireNonNull(original, "original is null");
+        retry = RetryDriver.retry()
+                .maxAttempts(5)
+                .stopRetryingWhen(e -> e instanceof ThriftServiceException && !((ThriftServiceException) e).isRetryPossible())
+                .exponentialBackoff(
+                        new Duration(10, TimeUnit.MILLISECONDS),
+                        new Duration(20, TimeUnit.MILLISECONDS),
+                        new Duration(30, TimeUnit.SECONDS),
+                        1.5);
+    }
+
+    @Override
+    public ThriftPrestoClient connectToAnyHost()
+    {
+        return new RetryingClient(original::connectToAnyHost, retry);
+    }
+
+    @Override
+    public ThriftPrestoClient connectToAnyOf(List<HostAddress> hosts)
+    {
+        return new RetryingClient(() -> original.connectToAnyOf(hosts), retry);
+    }
+
+    @NotThreadSafe
+    private static final class RetryingClient
+            implements ThriftPrestoClient
+    {
+        private final Supplier<ThriftPrestoClient> clientSupplier;
+        private final RetryDriver retry;
+        private ThriftPrestoClient client;
+
+        public RetryingClient(Supplier<ThriftPrestoClient> clientSupplier, RetryDriver retry)
+        {
+            this.clientSupplier = requireNonNull(clientSupplier, "clientSupplier is null");
+            this.retry = retry.onRetry(this::close);
+        }
+
+        private ThriftPrestoClient getClient()
+        {
+            if (client != null) {
+                return client;
+            }
+            else {
+                client = clientSupplier.get();
```

Comment:
`getClient()` is a low level method which is part of the code being retried on failure. Like this:
```
() -> getClient().listSchemaNames()
```